### PR TITLE
UISEES-30: Update filters options for Items with the facet data

### PR DESCRIPTION
**Pull request by Dmytro-Melnyshyn**
[UISEES-30](https://issues.folio.org/browse/UISEES-30): Update filters options for Items with the facet data

## Purpose

- create a custom hook with reusable facets functionality for all segments
- make changes to the `items` segment taking into account the custom hook functionality
- make changes to the `instances` segment

### CODE REVIEW

File: src/common/hooks/useFacets.js [hunk 1]
```diff
@@ -0,0 +1,447 @@
+import React, { useEffect, useRef, useState } from 'react';
+import { useLocation } from 'react-router-dom';
+import { FormattedMessage, useIntl } from 'react-intl';
+import _ from 'lodash';
+import { languageOptionsES } from '../../components/InstanceFilters/languages';
```
Comment (Godlevskyi) [inline src/common/hooks/useFacets.js]: Please add empty line here to separate global deps from stripes and local

File: src/common/hooks/useFacets.js [hunk 2]
```diff
@@ -0,0 +1,223 @@
+import { useEffect, useRef, useState } from 'react';
+import { useLocation } from 'react-router-dom';
+import _ from 'lodash';
+
+import { FACETS } from '../../constants';
+
+const useFacets = (segmentAccordions, segmentOptions, selectedFacetFilters, getNewRecords, data) => {
```
Comment (Godlevskyi) [inline src/common/hooks/useFacets.js]: Please format each prop on separate line

File: src/common/hooks/useFacets.js [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+      };
+    });
+
+    onFetchFacets({
+      accordions,
+      accordionsData: facetsData,
+    });
+  };
+
+  const handleFetchFacets = (property = {}) => {
```
Comment (Godlevskyi) [inline src/common/hooks/useFacets.js]: Please wrap this function in useCallback hook

File: src/common/hooks/useFacets.js [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+    setFacetSettings(prevFacetSettings => ({
+      ...prevFacetSettings,
+      [onMoreClickedFacet]: {
+        ...prevFacetSettings[onMoreClickedFacet],
+        isOnMoreClicked: true,
+      },
+    }));
+  };
+
+  const processAllFacets = () => {
```
Comment (Godlevskyi) [inline src/common/hooks/useFacets.js]: Please wrap it in useCallback hook

File: src/common/hooks/useFacets.js [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+    setFacetSettings(prevFacetSettings => ({
+      ...prevFacetSettings,
+      [name]: {
+        ...prevFacetSettings[name],
+        value,
+      },
+    }));
+  };
+
+  const processFilterChange = (selectedFilters, facetName) => {
```
Comment (Godlevskyi) [inline src/common/hooks/useFacets.js:58]: Please wrap it in useCallback hook

File: src/common/hooks/useFacets.js [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+  const [facetSettings, setFacetSettings] = useState({});
+  const [facetNameToOpen, setFacetNameToOpen] = useState('');
+  const [showLoadingForAllFacets, setShowLoadingForAllFacets] = useState(false);
+
+  const prevAccordionsState = useRef(accordions);
+  const prevFilters = useRef({});
+  const prevUrl = useRef({});
+  const prevQuery = useRef('');
+
+  const onToggleSection = ({ id }) => {
```
Comment (Godlevskyi) [inline src/common/hooks/useFacets.js]: Please wrap it in the useCallback

File: src/common/hooks/useFacets.js [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+    setAccordions(curState => {
+      const newState = _.cloneDeep(curState);
+      newState[id] = !curState[id];
+      return newState;
+    });
+  };
+
+  const handleFilterSearch = (filter) => {
```
Comment (Godlevskyi) [inline src/common/hooks/useFacets.js]: Please wrap it in the useCallback hook

File: src/common/hooks/useFacets.js [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+          [facetName]: {
+            ...prevAccordionsData[facetName],
+            isSelected: false,
+          },
+        }));
+      }
+    }
+  };
+
+  const processOnMoreClicking = (onMoreClickedFacet) => {
```
Comment (Godlevskyi) [inline src/common/hooks/useFacets.js]: Please wrap it in use callback hook

File: src/common/hooks/useFacets.js [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+    } else if (onMoreClickedFacet) {
+      processOnMoreClicking(onMoreClickedFacet);
+    } else if (focusedFacet) {
+      onFetchFacets({ focusedFacet });
+    } else {
+      processAllFacets();
+    }
+  };
+
+  const getIsPending = (facetName) => {
```
Comment (Godlevskyi) [inline src/common/hooks/useFacets.js]: Please wrap it in useCallback

File: src/routes/InstancesRoute.js [hunk 10]
```diff
@@ -15,10 +15,10 @@ import {
 } from './buildManifestObject';
 import { DataContext } from '../contexts';
 import {
-  FACETS_TO_REQUEST,
+  DEFAULT_FILTERS_NUMBER,
   FACETS,
-} from '../components/InstanceFilters/constants';
-import { DEFAULT_FILTERS_NUMBER } from '../constants';
+  FACETS_TO_REQUEST
+} from '../constants';
```
Comment (Godlevskyi) [inline src/routes/InstancesRoute.js:21]: Please delete useFacets empty file

File: src/common/hooks/useFacets.js (hunk 1 above)
Comment (Dmytro-Melnyshyn) [reply src/common/hooks/useFacets.js]: Thanks, fixed.

File: src/common/hooks/useFacets.js (hunk 2 above)
Comment (Dmytro-Melnyshyn) [reply src/common/hooks/useFacets.js]: Thanks, fixed.

File: src/common/hooks/useFacets.js (hunk 4 above)
Comment (Dmytro-Melnyshyn) [reply src/common/hooks/useFacets.js]: Thanks, fixed

File: src/common/hooks/useFacets.js (hunk 6 above)
Comment (Dmytro-Melnyshyn) [reply src/common/hooks/useFacets.js]: Thanks, fixed

File: src/common/hooks/useFacets.js (hunk 7 above)
Comment (Dmytro-Melnyshyn) [reply src/common/hooks/useFacets.js]: Thanks, fixed

File: src/common/hooks/useFacets.js (hunk 8 above)
Comment (Dmytro-Melnyshyn) [reply src/common/hooks/useFacets.js]: Thanks, fixed

File: src/common/hooks/useFacets.js (hunk 9 above)
Comment (Dmytro-Melnyshyn) [reply src/common/hooks/useFacets.js]: Thanks, fixed

File: src/routes/InstancesRoute.js (hunk 10 above)
Comment (Dmytro-Melnyshyn) [reply src/routes/InstancesRoute.js:21]: Thanks, fixed

File: src/common/hooks/useFacets.js (hunk 3 above)
Comment (Dmytro-Melnyshyn) [reply src/common/hooks/useFacets.js]: Thanks, fixed

File: src/common/hooks/useFacets.js (hunk 5 above)
Comment (Dmytro-Melnyshyn) [reply src/common/hooks/useFacets.js:58]: Thanks, fixed

**Review by Godlevskyi — APPROVED** [review]

**Pull request merged**